### DISCUSSION TITLE
Exception handler could not handle Exception but could handle HTTPExcetion

### DISCUSSION THREAD

**Discussion by shaoerkuai** [fastapi/fastapi#10832] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, WebSocket
from fastapi.exceptions import RequestValidationError
from fastapi.exceptions import HTTPException

app = FastAPI()

async def handle_http_exception(req, exc):
    logger.error(exc)
    return RestResponse.fail_internal(error=ErrorCode.ERROR_INTERNAL)

async def handle_exception(req, exc):
    return PlainTextResponse(str(exc), status_code=402)

app.add_exception_handler(HTTPException,handle_http_exception)
app.add_exception_handler(Exception,handle_exception)


router = APIRouter(prefix='/v1')


class NewModel:
    mes1: str
    mes2: str

    def __init__(self, var, var2):
        self.mes1 = var
        self.mes2 = var2


@router.get("/")
async def root():
    raise ValueError("xx")
    return RestResponse.fail(error=ErrorCode.ERROR_MAINTENANCE)


@router.get("/2")
async def root1():
    raise RequestValidationError('33')
    return RestResponse.success(data=NewModel('111', "sadsa"))

@router.get("/3")
async def root1():
    raise HTTPException(403)
    return RestResponse.success(data=NewModel('111', "sadsa"))
```

### Description

If I send request to api :/2 , /3 it will catch exception successfully like this
![image](https://github.com/tiangolo/fastapi/assets/117460332/f71bb51a-608d-49fb-8413-f5bf226b5a01)

But if I raise a ValueError and register a _builtin_ `Exception` handler, It's indeed returns customized reponse from `handle_exception` , but it seems that exception was not caught `[ERROR] 2023-12-26 01:34:31 | httptools_impl.run_asgi:409 | Exception in ASGI application`.

![image](https://github.com/tiangolo/fastapi/assets/117460332/2bee04e5-7e65-429d-85f7-02b6724006e1)

![image](https://github.com/tiangolo/fastapi/assets/117460332/6cc2b892-25b1-4572-8685-c7d4fd15fbd4)

I have searched some forums and it is mentioned that this can be achieved by overriding with custom API route:
```
class RouteErrorHandler(APIRoute):
    """Custom APIRoute that handles application errors and exceptions"""

    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            try:
                return await original_route_handler(request)
            except Exception as ex:
                if isinstance(ex, HTTPException):
                    raise ex
                logger.exception("uncaught error")
                # wrap error into pretty 500 exception
                raise HTTPException(status_code=500, detail=str(ex))

        return custom_route_handler


router = APIRouter(route_class=RouteErrorHandler)
```

 I have not tried if it is possible, but I want to confirm after some time, if this problem is on the latest FastAPI to solve?

### Operating System

Windows

### FastAPI Version

0.105.0

### Pydantic Version

1.10.2

### Python Version

Python 3.10.7

### Additional Context

I start the uvicorn app using:
```
if __name__ == '__main__':
    Config = uvicorn.Config(app='src.MyApp.App:app',
                            host='127.0.0.1',
                            port=8083,
                            reload=CONFIG_TOML['General']['Debug'],
                            headers=[("server", "")])
    AppLogger.logger_init()
    Server = uvicorn.Server(Config)
    Server.run()
```

**Comment by hasansezertasan** [comment]
Dear, provided example code doesn't seem to work.

I have cleaned a bit and imported some missing objects.

```python
from fastapi import FastAPI, APIRouter
from fastapi.responses import PlainTextResponse
from fastapi.exceptions import RequestValidationError
from fastapi.exceptions import HTTPException

app = FastAPI()

async def handle_http_exception(req, exc):
    print(exc)
    return PlainTextResponse(str(exc), status_code=402)

async def handle_exception(req, exc):
    return PlainTextResponse(str(exc), status_code=402)

app.add_exception_handler(HTTPException,handle_http_exception)
app.add_exception_handler(Exception,handle_exception)


router = APIRouter(prefix='/v1')


class NewModel:
    mes1: str
    mes2: str

    def __init__(self, var, var2):
        self.mes1 = var
        self.mes2 = var2


@router.get("/")
async def root():
    raise ValueError("xx")


@router.get("/2")
async def root1():
    raise RequestValidationError('33')

@router.get("/3")
async def root1():
    raise HTTPException(403)

```

Can you update this code block to re-generate the expected output?

**Comment by shaoerkuai** [comment]
Using your modified block with

api : (`"/"`) -> /v1:

```
[ERROR] 2023-12-26 22:14:38 | httptools_impl.run_asgi:409 | Exception in ASGI application

Traceback (most recent call last):

  File "D:\Project\fast-framework\backend\main.py", line 17, in <module>
    Server.run()
    │      └ <function Server.run at 0x00000237F23B7A30>
    └ <uvicorn.server.Server object at 0x00000237F2B631F0>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\uvicorn\server.py", line 60, in run
    return asyncio.run(self.serve(sockets=sockets))
           │       │   │    │             └ None
           │       │   │    └ <function Server.serve at 0x00000237F23B7AC0>
           │       │   └ <uvicorn.server.Server object at 0x00000237F2B631F0>
           │       └ <function run at 0x00000237F1FC2E60>
           └ <module 'asyncio' from 'C:\\Users\\86131\\AppData\\Local\\Programs\\Python\\Python310\\lib\\asyncio\\__init__.py'>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
           │    │                  └ <coroutine object Server.serve at 0x00000237F2B1B0D0>
           │    └ <function BaseEventLoop.run_until_complete at 0x00000237F1FD08B0>
           └ <_WindowsSelectorEventLoop running=True closed=False debug=False>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 633, in run_until_complete
    self.run_forever()
    │    └ <function BaseEventLoop.run_forever at 0x00000237F1FD0820>
    └ <_WindowsSelectorEventLoop running=True closed=False debug=False>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 600, in run_forever
    self._run_once()
    │    └ <function BaseEventLoop._run_once at 0x00000237F1FD2320>
    └ <_WindowsSelectorEventLoop running=True closed=False debug=False>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 1896, in _run_once
    handle._run()
    │      └ <function Handle._run at 0x00000237F1F71AB0>
    └ <Handle <TaskStepMethWrapper object at 0x00000237F2C3CBE0>()>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
    │    │            │    │           │    └ <member '_args' of 'Handle' objects>
    │    │            │    │           └ <Handle <TaskStepMethWrapper object at 0x00000237F2C3CBE0>()>
    │    │            │    └ <member '_callback' of 'Handle' objects>
    │    │            └ <Handle <TaskStepMethWrapper object at 0x00000237F2C3CBE0>()>
    │    └ <member '_context' of 'Handle' objects>
    └ <Handle <TaskStepMethWrapper object at 0x00000237F2C3CBE0>()>

> File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\uvicorn\protocols\http\httptools_impl.py", line 404, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
                   └ <uvicorn.middleware.proxy_headers.ProxyHeadersMiddleware object at 0x00000237F2B74C70>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
                 │    │   │      │        └ <bound method RequestResponseCycle.send of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x00000237F2...
                 │    │   │      └ <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x0000023...
                 │    │   └ {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('127.0.0.1', 8083), 'cl...
                 │    └ <fastapi.applications.FastAPI object at 0x00000237F2B75330>
                 └ <uvicorn.middleware.proxy_headers.ProxyHeadersMiddleware object at 0x00000237F2B74C70>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\fastapi\applications.py", line 1106, in __call__
    await super().__call__(scope, receive, send)
                           │      │        └ <bound method RequestResponseCycle.send of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x00000237F2...
                           │      └ <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x0000023...
                           └ {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('127.0.0.1', 8083), 'cl...

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
          │    │                │      │        └ <bound method RequestResponseCycle.send of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x00000237F2...
          │    │                │      └ <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x0000023...
          │    │                └ {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('127.0.0.1', 8083), 'cl...
          │    └ <starlette.middleware.errors.ServerErrorMiddleware object at 0x00000237F2C3C700>
          └ <fastapi.applications.FastAPI object at 0x00000237F2B75330>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
          │    │   │      │        └ <function ServerErrorMiddleware.__call__.<locals>._send at 0x00000237F2B86E60>
          │    │   │      └ <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x0000023...
          │    │   └ {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('127.0.0.1', 8083), 'cl...
          │    └ <starlette.middleware.exceptions.ExceptionMiddleware object at 0x00000237F2C3C6D0>
          └ <starlette.middleware.errors.ServerErrorMiddleware object at 0x00000237F2C3C700>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
          │    │   │      │        └ <function ExceptionMiddleware.__call__.<locals>.sender at 0x00000237F2C320E0>
          │    │   │      └ <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x0000023...
          │    │   └ {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('127.0.0.1', 8083), 'cl...
          │    └ <fastapi.middleware.asyncexitstack.AsyncExitStackMiddleware object at 0x00000237F2C3C5B0>
          └ <starlette.middleware.exceptions.ExceptionMiddleware object at 0x00000237F2C3C6D0>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\fastapi\middleware\asyncexitstack.py", line 20, in __call__
    raise e

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\fastapi\middleware\asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
          │    │   │      │        └ <function ExceptionMiddleware.__call__.<locals>.sender at 0x00000237F2C320E0>
          │    │   │      └ <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x0000023...
          │    │   └ {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('127.0.0.1', 8083), 'cl...
          │    └ <fastapi.routing.APIRouter object at 0x00000237F2BFAF80>
          └ <fastapi.middleware.asyncexitstack.AsyncExitStackMiddleware object at 0x00000237F2C3C5B0>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
          │     │      │      │        └ <function ExceptionMiddleware.__call__.<locals>.sender at 0x00000237F2C320E0>
          │     │      │      └ <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x0000023...
          │     │      └ {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('127.0.0.1', 8083), 'cl...
          │     └ <function Route.handle at 0x00000237F2ABC430>
          └ APIRoute(path='/v1/', name='root', methods=['GET'])

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
          │    │   │      │        └ <function ExceptionMiddleware.__call__.<locals>.sender at 0x00000237F2C320E0>
          │    │   │      └ <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x0000023...
          │    │   └ {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('127.0.0.1', 8083), 'cl...
          │    └ <function request_response.<locals>.app at 0x00000237F2C31CF0>
          └ APIRoute(path='/v1/', name='root', methods=['GET'])

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\routing.py", line 66, in app
    response = await func(request)
                     │    └ <starlette.requests.Request object at 0x00000237F2C3CE20>
                     └ <function get_request_handler.<locals>.app at 0x00000237F2C31C60>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\fastapi\routing.py", line 274, in app
    raw_response = await run_endpoint_function(
                         └ <function run_endpoint_function at 0x00000237F2A9EEF0>

  File "C:\Users\86131\AppData\Local\Programs\Python\Python310\lib\site-packages\fastapi\routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
                 │         │      └ {}
                 │         └ <function root at 0x00000237F2B244C0>
                 └ <fastapi.dependencies.models.Dependant object at 0x00000237F2BFAA40>

  File "D:\Project\fast-framework\backend\src\MyApp\api\v1_route.py", line 24, in root
    raise ValueError("xx")

ValueError: xx
```

Real HTTP output:
![image](https://github.com/tiangolo/fastapi/assets/117460332/9c3b27a3-0bad-4f3d-b8bb-4e1ee8a40559)

http output is custom handler response as it should be but still has another exception not caught which logged at loguru output.

api : (`"/2"`) -> /v1/2:

![image](https://github.com/tiangolo/fastapi/assets/117460332/e07918fe-cbc0-4d5b-8606-d3d3cbf748dd)


It works ok with my custom handler and no asgi exception.

api : (`"/3"`) -> /v1/3:

![image](https://github.com/tiangolo/fastapi/assets/117460332/2955addc-6882-406b-b01e-72fcecf65d3a)

It seems work too (with HTTPException string and handle_http_exception status code)

**Reply by hasansezertasan** [reply]
`/2` returns something like this:

```json
{
  "detail": "33"
}
```

It's because `RequestValidationError` is already added to the app as an exception handler.

You can override it by doing something like this to override the default behavior:

```python
app.add_exception_handler(RequestValidationError, custom_exception)

```

If that doesn't solve your problem, can you rephrase it by defining the behavior you get and the behavior you expect?

---

Here is my final code:

```python
from fastapi import FastAPI
from fastapi.responses import PlainTextResponse
from fastapi.exceptions import RequestValidationError
from fastapi.exceptions import HTTPException

app = FastAPI()

async def http_exception(req, exc):
    print("http_exception")
    return PlainTextResponse(str(exc), status_code=402)

async def all_exceptions(req, exc):
    print("all_exceptions")
    return PlainTextResponse(str(exc), status_code=402)

app.add_exception_handler(HTTPException, http_exception)
app.add_exception_handler(Exception, all_exceptions)
app.add_exception_handler(Exception, all_exceptions)


@app.get("/1")
async def root1():
    raise ValueError("xx")


@app.get("/2")
async def root2():
    raise RequestValidationError('33')

@app.get("/3")
async def root2():
    raise HTTPException(403)

```

**Comment by shaoerkuai** [comment]
I do succeed in handling the RequestValidationError, but I can't handle Python's lowest-level `Exception` properly, which comes from "/v1" instead of "/v1/2" or "/v1/3", `app.add_exception_handler(Exception, all_exceptions)`

I hope to catch the `lowest level Exception` without producing "Exception in ASGI application"

Just like  https://stackoverflow.com/questions/61596911/catch-exception-globally-in-fastapi mentioned.

**Comment by hasansezertasan** [comment]
Here is my latest code block:

```python
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from fastapi.exceptions import RequestValidationError
from fastapi.exceptions import HTTPException

app = FastAPI()

async def http_exception(
    req: Request,
    exc: Exception,
):
    print("http_exception")
    return JSONResponse(
        status_code=500,
        content={"message": "http_exception", "exc": str(exc),},
    )

async def request_validation_error(
    req: Request,
    exc: Exception,
):
    print("request_validation_error")
    return JSONResponse(
        status_code=500,
        content={"message": "request_validation_error", "exc": str(exc),},
    )

async def all_exceptions(
    req: Request,
    exc: Exception,
):
    print("all_exceptions")
    return JSONResponse(
        status_code=500,
        content={"message": "all_exceptions", "exc": str(exc),},
    )

app.add_exception_handler(Exception, all_exceptions)
app.add_exception_handler(HTTPException, http_exception)
app.add_exception_handler(RequestValidationError, request_validation_error)


@app.get("/all_exceptions")
async def root1():
    raise ZeroDivisionError("xx")


@app.get("/request_validation_error")
async def root2():
    raise RequestValidationError('33')

@app.get("/http_exception")
async def root3():
    raise HTTPException(403)

```

As you can see, I was able to catch `ZeroDivisionError("xx")` from `all_exceptions` handler and it returned this:

```json
{
  "message": "all_exceptions",
  "exc": "xx"
}
```

**Comment by shaoerkuai** [comment]
That's the quest is:
![image](https://github.com/tiangolo/fastapi/assets/117460332/c374f254-7849-4888-808c-09579a63c41a)

It returned
```
{
    "message": "all_exceptions",
    "exc": "xx"
}
```
 as expected but still raised `Exception in ASGI application` 

I don't know why this problem occurs, it seems that the exception is raised twice.
My expectation is that it will not raise `Exception in ASGI application` if I registerd a exception handler.

**Reply by hasansezertasan** [reply]
Well, it doesn't raise `Exception in ASGI application` if you specify the exception like this:

```python
app.add_exception_handler(ZeroDivisionError, all_exceptions)
```

And I believe you should bring this discussion to the Starlette Github Repo since everything with `exception_handlers` is handled by Starlette. FastAPI doesn't do much with those.

**Reply by hasansezertasan** [reply, marked as answer]
And I don't think the exception is raised twice. I believe it was logged because the exception was not handled by a specific handler, which I believe is a good thing.

In the end, you can always stop logging these, just configure the `uvicorn` logger.

**Comment by shaoerkuai** [comment]
Thanks, I will check `uvicorn` logger to disable that

**Reply by ghost** [reply]
I've encountered this issue too, how did you resolve it?

**Comment by Isuxiz** [comment]
I also encountered this problem. I have some custom exception classes, and a function to handle these exceptions, and register this function as an exception handler through `app.add_exception_handler`. The client successfully gets the response I set in the function, but I still see the traceback of the raised exception in the console.

**Comment by Isuxiz** [comment]
Maybe this will help: https://github.com/fastapi/fastapi/discussions/6376#discussioncomment-11085804
